File: examples/example_sverdrups_workflow.m
%% example_sverdrups_workflow.m
% End-to-end tour of the toolkit on data that is actually on disk.
%
% Sverdrups Glacier (NW Greenland) is the one place in the configured folders
% where a gridded observation subset and a redrawn flowline cover the same ice:
%   obs grid : SentinelMonthly_Sverdrups.nc  (72 monthly frames, 2015-2020)
%   flowline : sverdrups.mat                 (31 points, 0-7.5 km, 2015-2024)
%
% Demonstrates, in order:
%   1. the two observation loaders,
%   2. building an ROI from the flowline footprint, saving it (.mat + ISSM .exp)
%      and reading it back,
%   3. a speed map with the ROI and flowline overlaid,
%   4. moc_velocity_variance: melt-season window, two individual time slices,
%      and a robust record-long variant + coverage map,
%   5. a per-year table of relative change over the ROI,
%   6. the flowline: distance-time image and a point time series.
% The model section (7) is OFF by default: it loads a multi-GB ISSM .mat.
%
% Usage:
%   >> example_sverdrups_workflow
%   >> SAVE_FIGS = true;  example_sverdrups_workflow    % also write figs/*.png
%   >> RUN_MODEL = true;  example_sverdrups_workflow    % add the model panels

%% 0. Options ---------------------------------------------------------------
% Pre-set any of these before running to override the defaults.
if ~exist('OBS_NC','var'),     OBS_NC     = 'SentinelMonthly_Sverdrups.nc'; end
if ~exist('OBS_FL','var'),     OBS_FL     = 'sverdrups.mat';                end
if ~exist('SAVE_FIGS','var'),  SAVE_FIGS  = false;                          end
if ~exist('RUN_MODEL','var'),  RUN_MODEL  = false;                          end
if ~exist('MODEL_FILE','var'), MODEL_FILE = 'Model_NW_HindcastRun_TransientInversion.mat'; end

cfg = setup_moc();
figdir = fullfile(fileparts(fileparts(mfilename('fullpath'))), 'figs');
if SAVE_FIGS && ~isfolder(figdir), mkdir(figdir); end

%% 1. Load the observations -------------------------------------------------
O = moc_load_obs_netcdf(OBS_NC, 'bands',{'vv'});
F = moc_load_obs_flowline(OBS_FL);

fprintf('\nobs grid  : %dx%d px, %d frames %.2f-%.2f, %.0f%% of pixel-frames valid\n', ...
    numel(O.x), numel(O.y), numel(O.time), min(O.time), max(O.time), ...
    100*mean(isfinite(O.vv(:))));
fprintf('flowline  : %s, %d points (0-%.1f km), %d times %.2f-%.2f (source %s)\n', ...
    F.name, numel(F.d), max(F.d)/1e3, numel(F.time), min(F.time), max(F.time), F.source);

%% 2. ROI from the flowline footprint, saved and read back ------------------
pad = 3000;                                   % [m] around the flowline
xr = [min(F.x)-pad, max(F.x)+pad];
yr = [min(F.y)-pad, max(F.y)+pad];
roi = struct('name','sverdrups_trunk', 'kind','rectangle', 'epsg',cfg.epsg, ...
             'x',[xr(1) xr(2) xr(2) xr(1) xr(1)], ...
             'y',[yr(1) yr(1) yr(2) yr(2) yr(1)]);
moc_save_roi(roi);                            % rois/sverdrups_trunk.mat + .exp
roi = moc_load_roi('sverdrups_trunk');        % round-trip through the saved file

[Xo, Yo] = meshgrid(O.x, O.y);
fprintf('ROI       : %.1f x %.1f km, %d of %d grid pixels inside\n', ...
    diff(xr)/1e3, diff(yr)/1e3, sum(moc_roi_mask(roi, Xo, Yo), 'all'), numel(Xo));

%% 3. A speed map with the ROI and flowline on top --------------------------
[~, ti] = min(abs(O.time - 2019.6));          % late-summer frame
f1 = figure('Color','w','Position',[80 80 720 600]);
ax = moc_plot_map(O, 'band','vv', 'tindex',ti, 'roi',roi, 'clim',[0 3000], ...
                  'ax',axes());   % without 'ax' the plotters open their own figure
plot(ax, F.x, F.y, 'w-', 'LineWidth',2);
plot(ax, F.x, F.y, 'k--', 'LineWidth',1);
if SAVE_FIGS, exportgraphics(f1, fullfile(figdir,'01_speed_map.png'), 'Resolution',120); end

%% 4a. Melt-season variability, normalised ----------------------------------
% Peak-to-peak speed over May-Sep 2019, divided by each pixel's own mean speed
% over the same window: the left panel is comparable between neighbouring
% pixels, the right one is dominated by the fast trunk simply because it is fast.
% 'minspeed' matters here — drop it and the ROI mean jumps from ~0.35 to ~1.2 as
% near-stagnant margin pixels divide a few m/yr of noise by a few m/yr of speed.
V = moc_velocity_variance(O, roi, 'trange',[2019.35 2019.85], ...
                          'mincount',3, 'minspeed',20);

f2 = figure('Color','w','Position',[80 80 1200 520]);
moc_plot_variance(V, 'field','norm',   'ax',subplot(1,2,1), 'roi',roi, ...
                  'title','relative: range / mean speed [-]');
moc_plot_variance(V, 'field','spread', 'ax',subplot(1,2,2), 'roi',roi, ...
                  'title','absolute: range of speed [m/yr]');
if SAVE_FIGS, exportgraphics(f2, fullfile(figdir,'02_melt_season_variability.png'), 'Resolution',120); end

%% 4b. Between two individual time slices -----------------------------------
% Nearest frame to each date; this window also yields a SIGNED change field.
V2 = moc_velocity_variance(O, roi, 'times',[2019.15 2019.6]);

f3 = figure('Color','w','Position',[80 80 1200 520]);
moc_plot_variance(V2, 'field','normChange', 'ax',subplot(1,2,1), 'roi',roi, ...
                  'title',sprintf('relative change %.2f -> %.2f [-]', V2.time));
moc_plot_variance(V2, 'field','change',     'ax',subplot(1,2,2), 'roi',roi, ...
                  'title',sprintf('absolute change %.2f -> %.2f [m/yr]', V2.time));
if SAVE_FIGS, exportgraphics(f3, fullfile(figdir,'03_two_slice_change.png'), 'Resolution',120); end
fprintf('\nlate winter (%.2f) -> late summer (%.2f): mean |change| = %.0f%% of local speed\n', ...
    V2.time(1), V2.time(2), 100*V2.stats.normMean);

%% 4c. Robust, record-long variant + coverage -------------------------------
% IQR about the median is insensitive to a single bad frame; 'minspeed' keeps
% near-stagnant ice from producing enormous ratios, 'mincount' demands enough
% frames per pixel. The count map shows where that bites.
Vr = moc_velocity_variance(O, roi, 'statistic','iqr', 'normalize','median', ...
                           'minspeed',20, 'mincount',24);

f4 = figure('Color','w','Position',[80 80 1200 520]);
moc_plot_variance(Vr, 'field','norm',  'ax',subplot(1,2,1), 'roi',roi, ...
                  'title',sprintf('2015-2020 IQR / median speed [-]  (mean %.2f)', ...
                                  Vr.stats.normMean));
moc_plot_variance(Vr, 'field','count', 'ax',subplot(1,2,2), 'roi',roi, ...
                  'title','valid frames per pixel (of 72)');
if SAVE_FIGS, exportgraphics(f4, fullfile(figdir,'04_record_long_robust.png'), 'Resolution',120); end

%% 5. Per-year relative change over the ROI ---------------------------------
% evalc() just swallows the per-call summary line so the table stays readable.
years = 2015:2020;
fprintf('\n  melt season (Apr-Nov) over the ROI\n');
fprintf('  year | frames | px    | mean rel. | median rel. | mean range [m/yr]\n');
fprintf('  -----+--------+-------+-----------+-------------+------------------\n');
for yr = years
    try
        evalc(sprintf(['Vy = moc_velocity_variance(O, roi, ''trange'',[%f %f], ' ...
                       '''mincount'',3, ''minspeed'',20);'], yr+0.25, yr+0.85));
        fprintf('  %4d |  %5d | %5d |   %6.1f%% |     %6.1f%% |   %8.0f\n', ...
            yr, numel(Vy.time), Vy.stats.npix, 100*Vy.stats.normMean, ...
            100*Vy.stats.normMedian, Vy.stats.spreadMean);
    catch ME
        fprintf('  %4d |  skipped (%s)\n', yr, ME.message);
    end
end

%% 6. The flowline ----------------------------------------------------------
% Distance-time image of observed speed, and the series at one point. (The
% model-vs-obs versions of these are moc_plot_flowline / moc_plot_timeseries,
% exercised in section 7.)
f5 = figure('Color','w','Position',[80 80 1250 480]);

ax1 = subplot(1,2,1);
imagesc(ax1, F.time, F.d/1e3, F.vel, 'AlphaData', ~isnan(F.vel));
set(ax1,'YDir','normal'); axis(ax1,'tight');
colormap(ax1,'parula'); cb = colorbar(ax1); ylabel(cb,'speed [m/yr]');
xlabel(ax1,'decimal year'); ylabel(ax1,'along-flowline distance [km]');
title(ax1, sprintf('%s flowline (%s)', F.name, F.source), 'Interpreter','none');

[~, kp] = min(abs(F.d - 3000));               % ~3 km up-glacier
ax2 = subplot(1,2,2); hold(ax2,'on'); grid(ax2,'on');
if ~isempty(F.err)
    errorbar(ax2, F.time, F.vel(kp,:), F.err(kp,:), 'o', 'MarkerSize',3, ...
             'Color',[0.2 0.2 0.2], 'CapSize',0);
else
    plot(ax2, F.time, F.vel(kp,:), 'o-', 'MarkerSize',3, 'Color',[0.2 0.2 0.2]);
end
% The same pixel from the gridded product, for an independent check.
[~, cx] = min(abs(O.x - F.x(kp)));
[~, cy] = min(abs(O.y - F.y(kp)));
plot(ax2, O.time, squeeze(O.vv(cy,cx,:)), 's-', 'MarkerSize',3, ...
     'Color',[0.85 0.1 0.1], 'LineWidth',1);
legend(ax2, {'flowline file','gridded obs (nearest px)'}, 'Location','best');
xlabel(ax2,'decimal year'); ylabel(ax2,'speed [m/yr]');
title(ax2, sprintf('%s at d=%.1f km', F.name, F.d(kp)/1e3), 'Interpreter','none');
if SAVE_FIGS, exportgraphics(f5, fullfile(figdir,'05_flowline.png'), 'Resolution',120); end

%% 7. Model comparison (optional — loads a multi-GB ISSM .mat) --------------
if RUN_MODEL
    M = moc_load_model(MODEL_FILE, 'solution','transient', 'trange',[2018 2020]);

    D = moc_spatial_diff(M, O, roi, 'field','vel', 'obsband','vv', ...
                         'res',cfg.grid_res, 'time',2019.6, 'obswindow',0.05);
    f6 = moc_plot_diff(D, 'roi',roi);
    if SAVE_FIGS, exportgraphics(f6, fullfile(figdir,'06_model_obs_diff.png'), 'Resolution',120); end

    Fmod = moc_model_flowline(M, F);
    f7 = figure('Color','w','Position',[80 80 700 520]);
    moc_plot_flowline(Fmod, F, 'mode','profile', 'time',2019.6, 'ax',axes());
    if SAVE_FIGS, exportgraphics(f7, fullfile(figdir,'07_flowline_profile.png'), 'Resolution',120); end

    C = moc_compare_timeseries(M, F, 'dist',3000);
    f8 = figure('Color','w','Position',[80 80 800 480]);
    moc_plot_timeseries(C, 'ax',axes());
    if SAVE_FIGS, exportgraphics(f8, fullfile(figdir,'08_point_timeseries.png'), 'Resolution',120); end
else
    fprintf(['\n(model section skipped — set RUN_MODEL = true to load %s\n' ...
             ' and add the model/obs difference, flowline profile and paired series)\n'], ...
            MODEL_FILE);
end

if SAVE_FIGS
    fprintf('\nfigures written to %s\n', figdir);
end

File: python/examples/example_sverdrups_workflow.py
"""End-to-end tour of moc_py on the data that is actually on disk.

Python mirror of ``examples/example_sverdrups_workflow.m`` on the MATLAB side.
Sverdrups Glacier (NW Greenland) is the one place in the configured folders
where a gridded observation subset and a redrawn flowline cover the same ice:

    obs grid : SentinelMonthly_Sverdrups.nc  (72 monthly frames, 2015-2020)
    flowline : sverdrups.mat                 (31 points, 0-7.5 km, 2015-2024)
    model    : exported/Model_NW_fric1_Snapshot.nc  (376 steps, 2007-2022)

Demonstrates, in order:
    1. the two observation readers,
    2. an ROI built from the flowline footprint, saved (.json + ISSM .exp) and
       read back,
    3. a speed map with the ROI and flowline overlaid,
    4. :func:`moc_py.velocity_variance` — melt-season window, two individual
       time slices, and a robust record-long variant + coverage map,
    5. a per-year table of relative change over the ROI,
    6. the flowline: distance-time image and a point series checked against the
       nearest gridded pixel,
    7. the model: spatial difference, flowline profile, and a paired series.

Unlike the MATLAB tour, section 7 runs by default: the exported model netCDF
loads in ~1 s, where the MATLAB path would read a multi-GB ISSM ``.mat``.

Run::

    python examples/example_sverdrups_workflow.py                  # show figures
    python examples/example_sverdrups_workflow.py --save-figs      # write PNGs
    python examples/example_sverdrups_workflow.py --no-model       # obs only
"""

from __future__ import annotations

import argparse
import sys
from pathlib import Path

import numpy as np
import xarray as xr

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
import moc_py as mp  # noqa: E402

OBS_NC = "SentinelMonthly_Sverdrups.nc"
OBS_FL = "sverdrups.mat"
MODEL_NC = "exported/Model_NW_fric1_Snapshot.nc"

MELT_WINDOW = (2019.35, 2019.85)   # May-Sep 2019
TWO_SLICES = (2019.15, 2019.6)     # late winter -> late summer
POINT_DIST = 3000.0                # [m] along the flowline
MIN_SPEED = 20.0                   # [m/yr] below this a ratio is meaningless


def build_roi(F: xr.Dataset, pad: float = 3000.0,
              name: str = "sverdrups_trunk") -> mp.Roi:
    """Build, save and re-load a rectangular ROI around a flowline's footprint.

    Args:
        F: Flowline Dataset from :func:`moc_py.load_obs_flowline`.
        pad: Margin around the flowline bounding box [m].
        name: ROI name; also the saved filename stem in ``CONFIG.roi_dir``.

    Returns:
        The ROI, round-tripped through the saved ``.json`` (an ``.exp`` contour
        is written alongside it for the MATLAB/ISSM side).
    """
    x0, x1 = F.x.values.min() - pad, F.x.values.max() + pad
    y0, y1 = F.y.values.min() - pad, F.y.values.max() + pad
    roi = mp.Roi(name=name, kind="rectangle",
                 x=np.array([x0, x1, x1, x0, x0]),
                 y=np.array([y0, y0, y1, y1, y0]))
    mp.save_roi(roi)
    return mp.load_roi(name)


def yearly_table(O: xr.Dataset, roi: mp.Roi, years: range) -> None:
    """Print per-year relative change over the ROI for the melt season.

    Args:
        O: Gridded obs Dataset.
        roi: Region of interest.
        years: Years to tabulate (Apr-Nov of each is used).
    """
    print("\n  melt season (Apr-Nov) over the ROI")
    print("  year | frames | px    | mean rel. | median rel. | mean range [m/yr]")
    print("  -----+--------+-------+-----------+-------------+------------------")
    for yr in years:
        try:
            V = mp.velocity_variance(O, roi, trange=(yr + 0.25, yr + 0.85),
                                     min_count=3, min_speed=MIN_SPEED)
        except ValueError as e:
            print(f"  {yr:4d} |  skipped ({e})")
            continue
        a = V.attrs
        print(f"  {yr:4d} |  {a['nframes']:5d} | {a['npix']:5d} |"
              f"   {100 * a['norm_mean']:6.1f}% |     {100 * a['norm_median']:6.1f}% |"
              f"   {a['spread_mean']:8.0f}")


def flowline_figure(F: xr.Dataset, O: xr.Dataset, plt) -> "plt.Figure":
    """Draw the flowline distance-time image beside one point's time series.

    The gridded product is overlaid on the point series as an independent check
    that the two readers agree where they overlap.

    Args:
        F: Flowline Dataset.
        O: Gridded obs Dataset.
        plt: The pyplot module (passed in so the backend is chosen by the caller).

    Returns:
        The matplotlib ``Figure``.
    """
    fig, axs = plt.subplots(1, 2, figsize=(13, 4.8))
    mp.plot_flowline(F, mode="hovmoller", ax=axs[0])

    k = int(np.argmin(np.abs(F.d.values - POINT_DIST)))
    mp.plot_timeseries(F, point=k, ax=axs[1], use_datetime=False)
    ci = int(np.argmin(np.abs(O.x.values - float(F.x.values[k]))))
    ri = int(np.argmin(np.abs(O.y.values - float(F.y.values[k]))))
    axs[1].plot(O.time.values, O["vv"].isel(x=ci, y=ri).values, "s-", ms=3,
                color="#d81b1b", lw=1, label="gridded obs (nearest px)")
    axs[1].legend()
    fig.tight_layout()
    return fig


def model_figures(M, O: xr.Dataset, F: xr.Dataset, roi: mp.Roi, plt) -> list:
    """Model-vs-observation panels: spatial difference, profile, paired series.

    Args:
        M: Model from :func:`moc_py.load_model`.
        O: Gridded obs Dataset.
        F: Flowline Dataset.
        roi: Region of interest.
        plt: The pyplot module.

    Returns:
        List of ``(name, Figure)`` pairs.
    """
    D = mp.spatial_diff(M, O, roi, time=2019.6, obswindow=0.05)
    print(f"\nmodel - obs at t={D.attrs['time']:.2f} over the ROI: "
          f"bias={D.attrs['bias']:.0f}  RMSE={D.attrs['rmse']:.0f}  "
          f"MAE={D.attrs['mae']:.0f} m/yr over {D.attrs['n']} px")
    f6 = mp.plot_diff(D, roi=roi)

    Fmod = mp.model_flowline(M, F)
    f7, ax = plt.subplots(figsize=(7.5, 5))
    mp.plot_flowline(F, mode="profile", time=2019.6, model=Fmod, ax=ax)
    f7.tight_layout()

    C = mp.compare_timeseries(M, F, dist=POINT_DIST)
    print(f"at d={POINT_DIST / 1e3:.1f} km: bias={C.stats['mean']:.0f}  "
          f"RMSE={C.stats['rmse']:.0f} m/yr  r={C.stats['r']:.2f} "
          f"over {C.stats['n']} paired times")
    f8, ax = plt.subplots(figsize=(8, 4.8))
    mp.plot_compare(C, ax=ax, use_datetime=False)
    f8.tight_layout()

    return [("06_model_obs_diff", f6), ("07_flowline_profile", f7),
            ("08_point_timeseries", f8)]


def main(argv: list[str] | None = None) -> int:
    """Run the tour.

    Args:
        argv: Command-line arguments (default ``sys.argv[1:]``).

    Returns:
        Process exit code.
    """
    ap = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    ap.add_argument("--obs", default=OBS_NC, help="gridded obs netCDF")
    ap.add_argument("--flowline", default=OBS_FL, help="flowline .mat")
    ap.add_argument("--model", default=MODEL_NC, help="exported model netCDF")
    ap.add_argument("--no-model", action="store_true", help="skip section 7")
    ap.add_argument("--save-figs", action="store_true",
                    help="write PNGs to <repo>/figs/python instead of showing them")
    args = ap.parse_args(argv)

    import matplotlib
    if args.save_figs:
        matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    figs = []

    # --- 1. observations ----------------------------------------------------
    O = mp.load_obs_netcdf(args.obs, bands=["vv"])
    F = mp.load_obs_flowline(args.flowline)
    t = O.time.values
    print(f"\nobs grid  : {O.sizes['x']}x{O.sizes['y']} px, {t.size} frames "
          f"{t.min():.2f}-{t.max():.2f}, "
          f"{100 * np.isfinite(O['vv'].values).mean():.0f}% of pixel-frames valid")
    ft = F.time.values
    print(f"flowline  : {F.attrs.get('name', args.flowline)}, {F.sizes['point']} points "
          f"(0-{F.d.values.max() / 1e3:.1f} km), {ft.size} times "
          f"{ft.min():.2f}-{ft.max():.2f} (source {F.attrs.get('source', '?')})")

    # --- 2. ROI from the flowline footprint --------------------------------
    roi = build_roi(F)
    Xo, Yo = np.meshgrid(O.x.values, O.y.values)
    xmin, xmax, ymin, ymax = roi.bbox
    print(f"ROI       : {(xmax - xmin) / 1e3:.1f} x {(ymax - ymin) / 1e3:.1f} km, "
          f"{int(mp.roi_mask(roi, Xo, Yo).sum())} of {Xo.size} grid pixels inside")

    # --- 3. speed map with the ROI and flowline on top ---------------------
    ti = int(np.argmin(np.abs(t - 2019.6)))
    f1, ax = plt.subplots(figsize=(7.5, 6))
    mp.plot_map(O, band="vv", tindex=ti, ax=ax, clim=(0, 3000), roi=roi)
    ax.plot(F.x.values, F.y.values, "w-", lw=2.5)
    ax.plot(F.x.values, F.y.values, "k--", lw=1)
    f1.tight_layout()
    figs.append(("01_speed_map", f1))

    # --- 4a. melt-season variability, normalised ---------------------------
    # min_speed matters here: without it the ROI mean jumps from ~0.4 to ~1.2 as
    # near-stagnant margin pixels divide a few m/yr of noise by a few m/yr of speed.
    V = mp.velocity_variance(O, roi, trange=MELT_WINDOW,
                             min_count=3, min_speed=MIN_SPEED)
    f2, axs = plt.subplots(1, 2, figsize=(13, 5))
    mp.plot_variance(V, "norm", ax=axs[0], roi=roi,
                     title="relative: range / mean speed [-]")
    mp.plot_variance(V, "spread", ax=axs[1], roi=roi,
                     title="absolute: range of speed [m/yr]")
    f2.tight_layout()
    figs.append(("02_melt_season_variability", f2))

    # --- 4b. between two individual time slices ----------------------------
    V2 = mp.velocity_variance(O, roi, times=TWO_SLICES)
    t0, t1 = V2.time.values
    f3, axs = plt.subplots(1, 2, figsize=(13, 5))
    mp.plot_variance(V2, "norm_change", ax=axs[0], roi=roi,
                     title=f"relative change {t0:.2f} -> {t1:.2f} [-]")
    mp.plot_variance(V2, "change", ax=axs[1], roi=roi,
                     title=f"absolute change {t0:.2f} -> {t1:.2f} [m/yr]")
    f3.tight_layout()
    figs.append(("03_two_slice_change", f3))
    print(f"\nlate winter ({t0:.2f}) -> late summer ({t1:.2f}): "
          f"mean |change| = {100 * V2.attrs['norm_mean']:.0f}% of local speed")

    # --- 4c. robust, record-long variant + coverage ------------------------
    Vr = mp.velocity_variance(O, roi, statistic="iqr", normalize="median",
                              min_speed=MIN_SPEED, min_count=24)
    f4, axs = plt.subplots(1, 2, figsize=(13, 5))
    mp.plot_variance(Vr, "norm", ax=axs[0], roi=roi,
                     title=f"{t.min():.0f}-{t.max():.0f} IQR / median speed [-]"
                           f"  (mean {Vr.attrs['norm_mean']:.2f})")
    mp.plot_variance(Vr, "count", ax=axs[1], roi=roi,
                     title=f"valid frames per pixel (of {t.size})")
    f4.tight_layout()
    figs.append(("04_record_long_robust", f4))

    # --- 5. per-year table --------------------------------------------------
    yearly_table(O, roi, range(int(np.ceil(t.min())), int(np.floor(t.max())) + 1))

    # --- 6. the flowline ----------------------------------------------------
    figs.append(("05_flowline", flowline_figure(F, O, plt)))

    # --- 7. the model -------------------------------------------------------
    if args.no_model:
        print(f"\n(model section skipped — drop --no-model to load {args.model})")
    else:
        M = mp.load_model(args.model)
        print(f"\nmodel     : {M.x.size} vertices, {M.elements.shape[0]} elements, "
              f"{M.time.size} steps {M.time.min():.2f}-{M.time.max():.2f}")
        figs.extend(model_figures(M, O, F, roi, plt))

    # --- output -------------------------------------------------------------
    if args.save_figs:
        outdir = Path(__file__).resolve().parents[2] / "figs" / "python"
        outdir.mkdir(parents=True, exist_ok=True)
        for name, fig in figs:
            fig.savefig(outdir / f"{name}.png", dpi=120, bbox_inches="tight")
        print(f"\n{len(figs)} figures written to {outdir}")
    else:
        plt.show()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
